src/main/java/io/cryostat/platform/internal/DockerPlatformClient.java

```diff
@@ -0,0 +1,322 @@
+                                    "localhost",
+                                    requestPath.toString())
+                            .timeout(2_000L)
+                            .as(BodyCodec.string())
+                            .send(
+                                    ar -> {
+                                        if (ar.failed()) {
+                                            Throwable t = ar.cause();
+                                            logger.error("Docker API request failed", t);
+                                            result.completeExceptionally(t);
+                                            return;
+                                        }
+                                        result.complete(
+                                                gson.fromJson(
+                                                        ar.result().body(),
+                                                        ContainerDetails.class));
+                                    });
+                });
+        return result;
+    }
+
+    private ServiceRef convert(ContainerSpec desc) {
+        try {
+            JMXServiceURL connectUrl;
+            String hostname;
+            int jmxPort;
+            if (desc.Labels.containsKey(JMX_URL_LABEL)) {
+                connectUrl = new JMXServiceURL(desc.Labels.get(JMX_URL_LABEL));
+                if (URIUtil.isRmiUrl(connectUrl)) {
+                    URI serviceUrl = URIUtil.getRmiTarget(connectUrl);
+                    hostname = serviceUrl.getHost();
+                    jmxPort = serviceUrl.getPort();
+                } else {
+                    hostname = connectUrl.getHost();
+                    jmxPort = connectUrl.getPort();
+                }
+            } else {
+                jmxPort = Integer.parseInt(desc.Labels.get(JMX_PORT_LABEL));
+                hostname = desc.Labels.get(JMX_HOST_LABEL);
+                if (hostname == null) {
+                    if (desc.Names.size() < 1) {
+                        throw new IndexOutOfBoundsException();
+                    }
+                    hostname = desc.Names.get(0);
+                    if (hostname.startsWith("/")) {
+                        hostname = hostname.substring(1);
+                    }
+                }
+                connectUrl = connectionToolkit.get().createServiceURL(hostname, jmxPort);
+            }
+
+            Map<AnnotationKey, String> cryostatAnnotations = new HashMap<>();
+            cryostatAnnotations.put(AnnotationKey.REALM, REALM);
+
+            cryostatAnnotations.put(AnnotationKey.HOST, hostname);
+            cryostatAnnotations.put(AnnotationKey.PORT, Integer.toString(jmxPort));
+
+            ServiceRef serviceRef =
+                    new ServiceRef(
+                            null,
+                            URI.create(connectUrl.toString()),
+                            Optional.ofNullable(desc.Names.get(0)).orElse(desc.Id));
```

Comment:
`desc.Names.get(0)` here looks a lot like the `hostname` check a few lines higher. Maybe this can be refactored to use that existing value?